### PR TITLE
In filter_topo added extrapolation to halo points

### PR DESCRIPTION
**Pull request by RatkoVasic-NOAA**
This commit is related to the issue #105
Fields' geolat_c and geolon_c halo points are used to calculate dxc and dyc, but halo points are not initialized for regional standalone case. I added extrapolation for regional standalone similar way to regional nest and global application.

### CODE REVIEW

**Comment by GeorgeGayno-NOAA** [comment]
Thanks @RatkoVasic-NOAA. Does routine 'read_topo_file` have the same problem. What is in the halo region of arrays 'oro' and 'mask'?

**Comment by RatkoVasic-NOAA** [comment]
> Thanks @RatkoVasic-NOAA. Does routine 'read_topo_file` have the same problem. What is in the halo region of arrays 'oro' and 'mask'?

The only arrays (_in regional standalone mode_) that were using halo points are geolat_c and geolon_c. All other arrays are not using halo points.

File: sorc/grid_tools.fd/filter_topo.fd/filter_topo.F90 [hunk 1]
```diff
@@ -1144,6 +1147,42 @@ subroutine fill_cubic_grid_halo(data, data2, halo, ioff, joff, sign1, sign2)
 
   end subroutine fill_cubic_grid_halo
 
+  !> This routine extrapolate geolat_c and geolon_c halo points for the
+  !! regional standalone grid. Halo points are needed for dxc and dyc
+  !! calculation.
+  !!
```
Comment (GeorgeGayno-NOAA) [inline sorc/grid_tools.fd/filter_topo.fd/filter_topo.F90]: Replace the ??? with the variable definition. The ??? was a placeholder so we could get a basic doxygen capability in this code.  And use your full name and work center - "NCEP/EMC" - after 'author'.

File: sorc/grid_tools.fd/filter_topo.fd/filter_topo.F90 [hunk 2]
```diff
@@ -1144,6 +1147,42 @@ subroutine fill_cubic_grid_halo(data, data2, halo, ioff, joff, sign1, sign2)
...
+  !! @param[in] data  ???
+  !! @param[in] halo ???
+  !! @author Ratko
+  subroutine fill_regional_halo(data, halo)
```
Comment (GeorgeGayno-NOAA) [inline sorc/grid_tools.fd/filter_topo.fd/filter_topo.F90]: We are required to do unit tests for all new routines. These tests run under github actions. I can help you write the test. Just ask.

**Review by GeorgeGayno-NOAA — CHANGES_REQUESTED** [review]

**Comment by GeorgeGayno-NOAA** [comment]
@RatkoVasic-NOAA I tested your branch at b602411 on my canned test case. I compiled your branch on Hera with CMAKE_BUILD_TYPE equal to "Release" and "Debug". I got big differences along the halo.  I will provide details.

**Comment by GeorgeGayno-NOAA** [comment]
Here are the differences between "Debug" and "Release" mode for my test case:

```
+ nccmp -dmqfS oro.C424.tile7.nc oro.C424.tile7.nc.ratko.intel.O3
Variable  Group Count     Sum AbsSum     Min     Max   Range     Mean  StdDev
orog_filt /        55 1.90058 1901.6 -239.36 239.794 479.154 0.034556 70.2504
```

I visualized the files. The differences are confined to the halo.


If you do not already know, you can change the compile type using this variable: `-DCMAKE_BUILD_TYPE=Debug`

**Comment by RatkoVasic-NOAA** [comment]
I'll test it.

File: sorc/grid_tools.fd/filter_topo.fd/filter_topo.F90 (hunk 2 above)
Comment (RatkoVasic-NOAA) [reply sorc/grid_tools.fd/filter_topo.fd/filter_topo.F90]: I'm adding descriptions. I see everywhere @parm[in], does doxygen recognize [inout]?

Comment (RatkoVasic-NOAA) [reply sorc/grid_tools.fd/filter_topo.fd/filter_topo.F90]: Yes, I will need help to create test.

Comment (GeorgeGayno-NOAA) [reply sorc/grid_tools.fd/filter_topo.fd/filter_topo.F90]: > I'm adding descriptions. I see everywhere @parm[in], does doxygen recognize [inout]?

Yes.

Comment (RatkoVasic-NOAA) [reply sorc/grid_tools.fd/filter_topo.fd/filter_topo.F90]: > > I'm adding descriptions. I see everywhere @parm[in], does doxygen recognize [inout]?
> 
> Yes.

Oh, I see in manual, syntax is [in,out].

File: sorc/grid_tools.fd/filter_topo.fd/filter_topo.F90 [hunk 3]
```diff
@@ -1295,6 +1334,8 @@ subroutine two_delta_filter(is, ie, js, je, isd, ied, jsd, jed, npx, npy, ntiles
     integer:: i,j, nt, t
     integer:: is1, ie2, js1, je2
 
+    a2 = 0.
```
Comment (GeorgeGayno-NOAA) [inline sorc/grid_tools.fd/filter_topo.fd/filter_topo.F90]: @RatkoVasic-NOAA I might add a comment here about this being a simple fix and your suggestion for a better fix.

Comment (RatkoVasic-NOAA) [reply sorc/grid_tools.fd/filter_topo.fd/filter_topo.F90]: Let's hold on for the moment, I'm looking into better solution.

Comment (GeorgeGayno-NOAA) [reply sorc/grid_tools.fd/filter_topo.fd/filter_topo.F90]: Great. By the way, the random behavior seems to have disappeared. I tested 06cc530 using Intel/GNU, Release/Debug compilations. My canned case gave identical results for all four tests. So I think you found all the problem arrays.

**Comment by GeorgeGayno-NOAA** [comment]
@RatkoVasic-NOAA I tried 1ec6689 on my canned case. The intel/release and intel/debug options gave identical results. The gnu/release option gives a different answer from the intel options. And the gnu/release and gnu/debug options give different answers.  Differences are confined to the left edge of the grid. Here are the gnu differences:

```
$ nccmp -dmfqS oro.C424.tile7.nc.ratko.O3.gnu oro.C424.tile7.nc.ratko.debug.gnu
Variable  Group Count     Sum  AbsSum      Min     Max   Range    Mean  StdDev
orog_filt /        21 77.3102 97.5196 -7.94759 24.5957 32.5433 3.68144 7.90504
```

**Comment by RatkoVasic-NOAA** [comment]
I guess I'll have to work on this when I come back 😐. (Friday)Sent from my Galaxy
-------- Original message --------From: GeorgeGayno-NOAA ***@***.***> Date: 8/26/21  10:14 AM  (GMT-05:00) To: NOAA-EMC/UFS_UTILS ***@***.***> Cc: RatkoVasic-NOAA ***@***.***>, Mention ***@***.***> Subject: Re: [NOAA-EMC/UFS_UTILS] In filter_topo added extrapolation to halo points (#571) 
@RatkoVasic-NOAA I tried 1ec6689 on my canned case. The intel/release and intel/debug options gave identical results. The gnu/release option gives a different answer from the intel options. And the gnu/release and gnu/debug options give different answers.  Differences are confined to the left edge of the grid. Here are the gnu differences:
$ nccmp -dmfqS oro.C424.tile7.nc.ratko.O3.gnu oro.C424.tile7.nc.ratko.debug.gnu
Variable  Group Count     Sum  AbsSum      Min     Max   Range    Mean  StdDev
orog_filt /        21 77.3102 97.5196 -7.94759 24.5957 32.5433 3.68144 7.90504

**Comment by GeorgeGayno-NOAA** [comment]
Yes, this is a frustrating bug.

Also, I am writing a simple unit test for your new routine. I will check it in to your branch today.

**Comment by GeorgeGayno-NOAA** [comment]
@RatkoVasic-NOAA I added a unit test stub for you at f87f1f5.

**Comment by GeorgeGayno-NOAA** [comment]
@RatkoVasic-NOAA I see that your unit test was failing with a memory leak error. In your unit test, try deallocating "testdata" at the end of program execution.

**Comment by GeorgeGayno-NOAA** [comment]
It works!

**Comment by GeorgeGayno-NOAA** [comment]
@JeffBeck-NOAA may want to run a quick test with the regional workflow.

@RatkoVasic-NOAA is this ready to merge?

**Comment by RatkoVasic-NOAA** [comment]
> 
> @RatkoVasic-NOAA is this ready to merge?

I replied by email, I guess it didn't go through.
Answer is yes, it is ready. Is there anything I should do more for this PR?

**Comment by GeorgeGayno-NOAA** [comment]
> > @RatkoVasic-NOAA is this ready to merge?
> 
> I replied by email, I guess it didn't go through.
> Answer is yes, it is ready. Is there anything I should do more for this PR?

The 'grid_gen' regression tests will need to be run. Since results will likely change for the regional tests, new baseline files will need to be generated. @RatkoVasic-NOAA can you run the tests on Hera or WCOSS?

**Comment by GeorgeGayno-NOAA** [comment]
And merge to your branch the latest changes from 'develop'.

**Review by GeorgeGayno-NOAA — APPROVED** [review]

**Comment by GeorgeGayno-NOAA** [comment]
> > > @RatkoVasic-NOAA is this ready to merge?
> > 
> > 
> > I replied by email, I guess it didn't go through.
> > Answer is yes, it is ready. Is there anything I should do more for this PR?
> 
> The 'grid_gen' regression tests will need to be run. Since results will likely change for the regional tests, new baseline files will need to be generated. @RatkoVasic-NOAA can you run the tests on Hera or WCOSS?

Thanks. I have updated the baseline data on all machines.

**Pull request merged**